### PR TITLE
Source Google Analytics Data API: ability to add multiple property ids

### LINKED ISSUE #24739: Source Google Analytics 4 (GA4): Add the ability to select multiple property id when setting up a GA4 connector
## Tell us about the problem you're trying to solve
To get all the data about all our sites, I need to add 100+ GA4 property IDs, which means at this point I need to create 100+ threads with similar queries.

## Describe the solution you’d like
Allow the GA4 connector to accept and process my property identifiers in comma or other format (json, array, etc.) when configured.

## Describe the alternative you’ve considered or used
Creating 100+ threads with the same query parameters.

## Are you willing to submit a PR?
Yes, but as soon as we have votes for this work to prioritize.

### PR DESCRIPTION
**Pull request by roman-yermilov-gl**
## What
*https://github.com/airbytehq/airbyte/issues/24739*

## How
*Property ID is now a list. Set of streams is generated based on each property id*

### CODE REVIEW

**Comment by roman-yermilov-gl** [comment]
There is another option that let us not to rollback to old API version. We can add `?include=customer` query param: (https://developer.rechargepayments.com/2021-11/extending_responses), so we will have entier `customer` model included in response in this way:
```
{
    "id": 534919106,
    ...,
    "customer": {
        "id": 64962974,
        "email": "email@example.com",
        "external_customer_id": {
            "ecommerce": "5213433266347"
        },
        "hash": "f99bd4a6877257af"
    },
    "include": {
        "customer": {
            "id": 10000000,
            "analytics_data": {
                "utm_params": []
            },
            "created_at": "2021-05-13T09:46:44+00:00",
            "email": "email@example.com",
            "external_customer_id": {
                "ecommerce": "5213433266347"
            },
            "first_charge_processed_at": "2021-05-13T13:46:39+00:00",
            "first_name": "Banana",
            "has_payment_method_in_dunning": false,
            "has_valid_payment_method": true,
            "hash": "f99bd4a6877257af",
            "last_name": "Green",
            "phone": "+11111111111",
            "subscriptions_active_count": 1,
            "subscriptions_total_count": 1,
            "tax_exempt": false,
            "updated_at": "2023-05-13T04:16:36+00:00"
        }
    }
    ...
}
```
It let us transform a record, so we can put missing fields from `include.customer.<field>` to `customer.<field>`. There are also some disadvatages: record will contain more data and we will have some duplicated fields.
I am not requesting change. This is only an another of how we can fix this problem

File: airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_source.py [hunk 1]
```diff
@@ -43,7 +43,7 @@ def patch_base_class():
 @pytest.fixture
 def config():
     return {
-        "property_id": "108176369",
+        "property_ids": ["108176369"],
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_source.py:46]: can we add unit test with old config and try test backward compatibility in that way

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/spec.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
     "$schema": "https://json-schema.org/draft-07/schema#",
     "title": "Google Analytics (Data API) Spec",
     "type": "object",
-    "required": ["property_id", "date_ranges_start_date"],
+    "required": ["property_ids", "date_ranges_start_date"],
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/spec.json:7]: do you test those changes manually for backward compatibility?

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py [hunk 3]
```diff
@@ -412,68 +412,81 @@ def get_authenticator(self, config: Mapping[str, Any]):
...
+
+            if not metadata:
+                return False, "failed to get metadata, over quota, try later"
+
+            dimensions = {d["apiName"] for d in metadata["dimensions"]}
+            metrics = {d["apiName"] for d in metadata["metrics"]}
+
+            for report in _config["custom_reports"]:
+                invalid_dimensions = set(report["dimensions"]) - dimensions
+                if invalid_dimensions:
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py:460]: please add spaces between condition blocks and add comments for them

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py [hunk 4]
```diff
@@ -412,68 +412,81 @@ def get_authenticator(self, config: Mapping[str, Any]):
...
+                    )
+                    message = f"Access was denied to the property ID entered. Check your access to the Property ID or use Google Analytics {property_id_docs_url} to find your Property ID."
+
+                    wrong_property_id_error = AirbyteTracedException(
+                        message=message, internal_message=internal_message, failure_type=FailureType.config_error
+                    )
+                    raise wrong_property_id_error
+
+            if not metadata:
+                return False, "failed to get metadata, over quota, try later"
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py:453]: Error message better start with upper case letter

File: airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py [hunk 5]
```diff
@@ -39,6 +39,7 @@ def patch_base_class(mocker):
 
     return {
         "config": {
+            "property_ids": ["496180525"],
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py:42]: if an old config has not property_ids 

what happens next?

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py (hunk 4 above)
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py]: Done

**Review by lazebnyi — CHANGES_REQUESTED** [review]
Please run tests with old format config

**Comment by roman-yermilov-gl** [comment]
> Please run tests with old format config

The tests were run with old config. If format is new, it means migration was applied during CI run

**Review by lazebnyi — APPROVED** [review]

**Review by lazebnyi — APPROVED** [review]

**Comment by Talalqaddoumi** [comment]
I recommend consolidating the data into the same stream rather than creating a stream per table per property. The latter approach proves unhelpful and leads to complications when refreshing the source schema, especially after the addition of numerous properties. Currently, I'm facing difficulties generating and refreshing the source schema in a setup with 160 properties.

Here is an illustration of the challenge:
image

Furthermore, adopting this approach would necessitate additional transformations in the destination to append the data into a single table.

**Pull request merged**